spirv_glsl.cpp

```diff
@@ -3258,6 +3306,78 @@ void CompilerGLSL::emit_resources()
 	declare_undefined_values();
 }
 
+void CompilerGLSL::emit_extension_workarounds()
+{
+	//GL_KHR_shader_subgroup_vote
+	if (!options.vulkan_semantics && declaredKHR_shader_subgroup_vote)
+	{
+		statement("#ifdef GL_ARB_shader_group_vote");
+		statement("bool subgroupAll(in bool v) { return allInvocationsARB(v); }");
+		statement("bool subgroupAny(in bool v) { return anyInvocationARB(v); }");
+		statement("#define SUBGROUP_ALL_EQUAL_WORKAROUND(type) bool subgroupAllEqual(type value) { return "
+		          "allInvocationsEqualARB(equal(value)); }");
+		statement("SUBGROUP_ALL_EQUAL_WORKAROUND(int)");
+		statement("SUBGROUP_ALL_EQUAL_WORKAROUND(ivec2)");
+		statement("SUBGROUP_ALL_EQUAL_WORKAROUND(ivec3)");
+		statement("SUBGROUP_ALL_EQUAL_WORKAROUND(ivec4)");
+		statement("SUBGROUP_ALL_EQUAL_WORKAROUND(uint)");
+		statement("SUBGROUP_ALL_EQUAL_WORKAROUND(uvec2)");
+		statement("SUBGROUP_ALL_EQUAL_WORKAROUND(uvec3)");
+		statement("SUBGROUP_ALL_EQUAL_WORKAROUND(uvec4)");
+		statement("SUBGROUP_ALL_EQUAL_WORKAROUND(float)");
+		statement("SUBGROUP_ALL_EQUAL_WORKAROUND(vec2)");
+		statement("SUBGROUP_ALL_EQUAL_WORKAROUND(vec3)");
+		statement("SUBGROUP_ALL_EQUAL_WORKAROUND(vec4)");
+		statement("SUBGROUP_ALL_EQUAL_WORKAROUND(double)");
+		statement("SUBGROUP_ALL_EQUAL_WORKAROUND(dvec2)");
+		statement("SUBGROUP_ALL_EQUAL_WORKAROUND(dvec3)");
+		statement("SUBGROUP_ALL_EQUAL_WORKAROUND(dvec4)");
+		statement("#undef SUBGROUP_ALL_EQUAL_WORKAROUND");
```

Comment:
To be able to service these versions of `subgroupAllEqual`, you'd need to implicitly enable `GL_KHR_shader_subgroup_ballot` (which you should only do if you detect the SPIR-V Opcodes for `subgroupAllEqual` present in the source)

Reply:
```glsl
bool subgroupAllEqual(NotBoolType value)
{
    return subgroupAllEqual(subgroupBroadcast(value,subgroupElect(void))==value);
}
```
That's what the macro should look like.

Reply:
Also define and use the macro later, its the same workaround, regardless of the ARB/NV/AMD extension used to emulate.

Reply:
`subgroupBroadcast(value, subgroupElect(void))`, shouldn't that be `subgroupBroadcastFirst(value) == value`?

Reply:
yeah you're right about the  `subgroupBroadcastFirst`, edited.